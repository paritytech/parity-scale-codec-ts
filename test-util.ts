/// <reference lib="deno.unstable"/>

import { dirname, join } from "std/path/win32.ts";
import { assertEquals } from "std/testing/asserts.ts";
import { assertSnapshot } from "std/testing/snapshot.ts";
import { Codec } from "./common.ts";

const file = (path: string) => () => fetch(join(dirname(import.meta.url), path)).then((x) => x.text());
export const files = {
  lipsum: file("lipsum.txt"),
  words: file("words.txt"),
  cargoLock: file("Cargo.lock"),
};

export function testCodec<T>(name: string, codec: Codec<T>, values: T[] | Record<string, T>) {
  for (const key in values) {
    const value = values[key as never] as T;
    const label = values instanceof Array
      ? Deno.inspect(value, {
        depth: Infinity,
        trailingComma: true,
        compact: true,
        iterableLimit: Infinity,
      })
      : key;
    Deno.test(`${name} ${label}`, async (t) => {
      const encoded = codec.encode(value);
      const decoded = codec.decode(encoded);
      assertEquals(decoded, value);
      await assertSnapshot(t, encoded, { serializer: serializeU8A });
    });
  }
}

function serializeU8A(array: Uint8Array) {
  return [...array].map((x) => x.toString(16).padStart(2, "0")).join("\n");
}

export function benchCodec<T>(name: string, codec: Codec<T>, value: T) {
  const encoded = codec.encode(value);
  Deno.bench(`- ${name} (${encoded.length}B) [encode] `, () => {
    codec.encode(value);
  });
  Deno.bench(`  ${name} (${encoded.length}B) [decode]`, () => {
    codec.decode(encoded);
  });
}
