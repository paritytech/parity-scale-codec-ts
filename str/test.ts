import * as asserts from "std/testing/asserts.ts";
import * as s from "../mod.ts";
import * as f from "../test-util.ts";

Deno.test("Strings", () => {
  f.visitFixtures(f.fixtures.str_, (bytes, decoded) => {
    asserts.assertEquals(s.str.decode(bytes), decoded);
    asserts.assertEquals(s.str.encode(decoded), bytes);
  }, f.constrainedIdentity<string>());
});
