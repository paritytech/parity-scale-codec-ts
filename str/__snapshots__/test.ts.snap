export const snapshot = {};

snapshot[`str "" 1`] = `00`;

snapshot[`str quickBrownFox 1`] = `
ac
54
68
65
20
71
75
69
63
6b
20
62
72
6f
77
6e
20
66
6f
78
20
6a
75
6d
70
73
20
6f
76
65
72
20
74
68
65
20
6c
61
7a
79
20
64
6f
67
`;

snapshot[`str lipsum 1`] = `
f9
06
4c
6f
72
65
6d
20
69
70
73
75
6d
20
64
6f
6c
6f
72
20
73
69
74
20
61
6d
65
74
2c
20
63
6f
6e
73
65
63
74
65
74
75
72
20
61
64
69
70
69
73
63
69
6e
67
20
65
6c
69
74
2c
20
73
65
64
20
64
6f
20
65
69
75
73
6d
6f
64
20
74
65
6d
70
6f
72
20
69
6e
63
69
64
69
64
75
6e
74
20
75
74
20
6c
61
62
6f
72
65
20
65
74
20
64
6f
6c
6f
72
65
20
6d
61
67
6e
61
20
61
6c
69
71
75
61
2e
20
55
74
20
65
6e
69
6d
20
61
64
20
6d
69
6e
69
6d
20
76
65
6e
69
61
6d
2c
20
71
75
69
73
20
6e
6f
73
74
72
75
64
20
65
78
65
72
63
69
74
61
74
69
6f
6e
20
75
6c
6c
61
6d
63
6f
20
6c
61
62
6f
72
69
73
20
6e
69
73
69
20
75
74
20
61
6c
69
71
75
69
70
20
65
78
20
65
61
20
63
6f
6d
6d
6f
64
6f
20
63
6f
6e
73
65
71
75
61
74
2e
20
44
75
69
73
20
61
75
74
65
20
69
72
75
72
65
20
64
6f
6c
6f
72
20
69
6e
20
72
65
70
72
65
68
65
6e
64
65
72
69
74
20
69
6e
20
76
6f
6c
75
70
74
61
74
65
20
76
65
6c
69
74
20
65
73
73
65
20
63
69
6c
6c
75
6d
20
64
6f
6c
6f
72
65
20
65
75
20
66
75
67
69
61
74
20
6e
75
6c
6c
61
20
70
61
72
69
61
74
75
72
2e
20
45
78
63
65
70
74
65
75
72
20
73
69
6e
74
20
6f
63
63
61
65
63
61
74
20
63
75
70
69
64
61
74
61
74
20
6e
6f
6e
20
70
72
6f
69
64
65
6e
74
2c
20
73
75
6e
74
20
69
6e
20
63
75
6c
70
61
20
71
75
69
20
6f
66
66
69
63
69
61
20
64
65
73
65
72
75
6e
74
20
6d
6f
6c
6c
69
74
20
61
6e
69
6d
20
69
64
20
65
73
74
20
6c
61
62
6f
72
75
6d
2e
0a
`;

snapshot[`str words 1`] = `
c5
02
62
69
6e
64
67
65
6e
0a
70
61
72
69
74
79
74
65
63
68
0a
6c
69
70
73
75
6d
0a
63
75
6d
6d
6f
6e
0a
62
69
74
76
65
63
0a
78
66
66
6e
0a
64
70
72
69
6e
74
0a
73
6f
72
61
6d
69
74
73
75
0a
68
79
64
72
6f
6b
69
6e
65
73
69
73
0a
6d
61
67
7a
0a
64
65
6e
6f
0a
63
6f
6d
62
69
6e
61
74
6f
72
73
0a
74
79
70
65
6f
66
0a
64
65
6e
6f
6c
61
6e
64
0a
63
6f
6e
74
72
61
76
61
72
69
61
6e
63
65
0a
72
75
73
74
75
70
0a
4c
69
6f
70
6c
65
75
72
6f
64
6f
6e
0a
4b
6f
73
6d
6f
63
65
72
61
74
6f
70
73
0a
50
73
69
74
74
61
63
6f
73
61
75
72
75
73
0a
`;

snapshot[`str cargoLock 1`] = `
5d
4d
23
20
54
68
69
73
20
66
69
6c
65
20
69
73
20
61
75
74
6f
6d
61
74
69
63
61
6c
6c
79
20
40
67
65
6e
65
72
61
74
65
64
20
62
79
20
43
61
72
67
6f
2e
0a
23
20
49
74
20
69
73
20
6e
6f
74
20
69
6e
74
65
6e
64
65
64
20
66
6f
72
20
6d
61
6e
75
61
6c
20
65
64
69
74
69
6e
67
2e
0a
76
65
72
73
69
6f
6e
20
3d
20
33
0a
0a
5b
5b
70
61
63
6b
61
67
65
5d
5d
0a
6e
61
6d
65
20
3d
20
22
61
72
72
61
79
76
65
63
22
0a
76
65
72
73
69
6f
6e
20
3d
20
22
30
2e
37
2e
32
22
0a
73
6f
75
72
63
65
20
3d
20
22
72
65
67
69
73
74
72
79
2b
68
74
74
70
73
3a
2f
2f
67
69
74
68
75
62
2e
63
6f
6d
2f
72
75
73
74
2d
6c
61
6e
67
2f
63
72
61
74
65
73
2e
69
6f
2d
69
6e
64
65
78
22
0a
63
68
65
63
6b
73
75
6d
20
3d
20
22
38
64
61
35
32
64
36
36
63
37
30
37
31
65
32
65
33
66
61
32
61
31
65
35
63
36
64
30
38
38
66
65
63
34
37
62
35
39
33
30
33
32
62
32
35
34
66
35
65
39
38
30
64
65
38
65
61
35
34
34
35
34
64
36
22
0a
0a
5b
5b
70
61
63
6b
61
67
65
5d
5d
0a
6e
61
6d
65
20
3d
20
22
62
69
74
76
65
63
22
0a
76
65
72
73
69
6f
6e
20
3d
20
22
31
2e
30
2e
30
22
0a
73
6f
75
72
63
65
20
3d
20
22
72
65
67
69
73
74
72
79
2b
68
74
74
70
73
3a
2f
2f
67
69
74
68
75
62
2e
63
6f
6d
2f
72
75
73
74
2d
6c
61
6e
67
2f
63
72
61
74
65
73
2e
69
6f
2d
69
6e
64
65
78
22
0a
63
68
65
63
6b
73
75
6d
20
3d
20
22
31
34
38
39
66
63
62
39
33
61
35
62
62
34
37
64
61
30
34
36
32
63
61
39
33
61
64
32
35
32
61
64
36
61
66
32
31
34
35
63
63
65
35
38
64
31
30
64
34
36
61
38
33
39
33
31
62
61
39
66
30
31
36
62
22
0a
64
65
70
65
6e
64
65
6e
63
69
65
73
20
3d
20
5b
0a
20
22
66
75
6e
74
79
22
2c
0a
20
22
72
61
64
69
75
6d
22
2c
0a
20
22
74
61
70
22
2c
0a
20
22
77
79
7a
22
2c
0a
5d
0a
0a
5b
5b
70
61
63
6b
61
67
65
5d
5d
0a
6e
61
6d
65
20
3d
20
22
62
79
74
65
2d
73
6c
69
63
65
2d
63
61
73
74
22
0a
76
65
72
73
69
6f
6e
20
3d
20
22
31
2e
32
2e
31
22
0a
73
6f
75
72
63
65
20
3d
20
22
72
65
67
69
73
74
72
79
2b
68
74
74
70
73
3a
2f
2f
67
69
74
68
75
62
2e
63
6f
6d
2f
72
75
73
74
2d
6c
61
6e
67
2f
63
72
61
74
65
73
2e
69
6f
2d
69
6e
64
65
78
22
0a
63
68
65
63
6b
73
75
6d
20
3d
20
22
38
37
63
35
66
64
64
30
31
36
36
30
39
35
65
31
64
34
36
33
66
63
36
63
63
30
31
61
61
38
63
65
35
34
37
61
64
37
37
61
34
65
38
34
64
34
32
65
62
36
37
36
32
62
30
38
34
65
32
38
30
36
37
65
22
0a
0a
5b
5b
70
61
63
6b
61
67
65
5d
5d
0a
6e
61
6d
65
20
3d
20
22
65
69
74
68
65
72
22
0a
76
65
72
73
69
6f
6e
20
3d
20
22
31
2e
36
2e
31
22
0a
73
6f
75
72
63
65
20
3d
20
22
72
65
67
69
73
74
72
79
2b
68
74
74
70
73
3a
2f
2f
67
69
74
68
75
62
2e
63
6f
6d
2f
72
75
73
74
2d
6c
61
6e
67
2f
63
72
61
74
65
73
2e
69
6f
2d
69
6e
64
65
78
22
0a
63
68
65
63
6b
73
75
6d
20
3d
20
22
65
37
38
64
34
66
31
63
63
34
61
65
33
33
62
62
66
63
31
35
37
65
64
35
64
35
61
35
65
66
33
62
63
32
39
32
32
37
33
30
33
64
35
39
35
38
36
31
64
65
62
32
33
38
66
63
65
63
34
65
39
34
35
37
22
0a
0a
5b
5b
70
61
63
6b
61
67
65
5d
5d
0a
6e
61
6d
65
20
3d
20
22
66
75
6e
74
79
22
0a
76
65
72
73
69
6f
6e
20
3d
20
22
32
2e
30
2e
30
22
0a
73
6f
75
72
63
65
20
3d
20
22
72
65
67
69
73
74
72
79
2b
68
74
74
70
73
3a
2f
2f
67
69
74
68
75
62
2e
63
6f
6d
2f
72
75
73
74
2d
6c
61
6e
67
2f
63
72
61
74
65
73
2e
69
6f
2d
69
6e
64
65
78
22
0a
63
68
65
63
6b
73
75
6d
20
3d
20
22
65
36
64
35
61
33
32
38
31
35
61
65
33
66
33
33
33
30
32
64
39
35
66
64
63
62
32
63
65
31
37
38
36
32
66
38
63
36
35
33
36
33
64
63
66
64
32
39
33
36
30
34
38
30
62
61
31
30
30
31
66
63
39
63
22
0a
0a
5b
5b
70
61
63
6b
61
67
65
5d
5d
0a
6e
61
6d
65
20
3d
20
22
69
6d
70
6c
2d
74
72
61
69
74
2d
66
6f
72
2d
74
75
70
6c
65
73
22
0a
76
65
72
73
69
6f
6e
20
3d
20
22
30
2e
32
2e
32
22
0a
73
6f
75
72
63
65
20
3d
20
22
72
65
67
69
73
74
72
79
2b
68
74
74
70
73
3a
2f
2f
67
69
74
68
75
62
2e
63
6f
6d
2f
72
75
73
74
2d
6c
61
6e
67
2f
63
72
61
74
65
73
2e
69
6f
2d
69
6e
64
65
78
22
0a
63
68
65
63
6b
73
75
6d
20
3d
20
22
31
31
64
37
61
39
66
36
33
33
30
62
37
31
66
65
61
35
37
39
32
31
63
39
62
36
31
63
34
37
65
65
36
65
38
34
66
37
32
64
33
39
34
37
35
34
65
66
66
36
31
36
33
61
65
36
37
65
37
33
39
35
65
62
22
0a
64
65
70
65
6e
64
65
6e
63
69
65
73
20
3d
20
5b
0a
20
22
70
72
6f
63
2d
6d
61
63
72
6f
32
22
2c
0a
20
22
71
75
6f
74
65
22
2c
0a
20
22
73
79
6e
22
2c
0a
5d
0a
0a
5b
5b
70
61
63
6b
61
67
65
5d
5d
0a
6e
61
6d
65
20
3d
20
22
69
74
65
72
74
6f
6f
6c
73
22
0a
76
65
72
73
69
6f
6e
20
3d
20
22
30
2e
31
30
2e
33
22
0a
73
6f
75
72
63
65
20
3d
20
22
72
65
67
69
73
74
72
79
2b
68
74
74
70
73
3a
2f
2f
67
69
74
68
75
62
2e
63
6f
6d
2f
72
75
73
74
2d
6c
61
6e
67
2f
63
72
61
74
65
73
2e
69
6f
2d
69
6e
64
65
78
22
0a
63
68
65
63
6b
73
75
6d
20
3d
20
22
61
39
61
39
64
31
39
66
61
31
65
37
39
62
36
32
31
35
66
66
32
39
62
39
64
36
38
38
30
62
37
30
36
31
34
37
66
31
36
65
39
62
31
64
62
62
31
65
34
65
35
39
34
37
62
35
62
30
32
62
63
35
65
33
22
0a
64
65
70
65
6e
64
65
6e
63
69
65
73
20
3d
20
5b
0a
20
22
65
69
74
68
65
72
22
2c
0a
5d
0a
0a
5b
5b
70
61
63
6b
61
67
65
5d
5d
0a
6e
61
6d
65
20
3d
20
22
70
61
72
69
74
79
2d
73
63
61
6c
65
2d
63
6f
64
65
63
22
0a
76
65
72
73
69
6f
6e
20
3d
20
22
33
2e
31
2e
32
22
0a
73
6f
75
72
63
65
20
3d
20
22
72
65
67
69
73
74
72
79
2b
68
74
74
70
73
3a
2f
2f
67
69
74
68
75
62
2e
63
6f
6d
2f
72
75
73
74
2d
6c
61
6e
67
2f
63
72
61
74
65
73
2e
69
6f
2d
69
6e
64
65
78
22
0a
63
68
65
63
6b
73
75
6d
20
3d
20
22
65
38
62
34
34
34
36
31
36
33
35
62
62
62
31
61
30
33
30
30
66
31
30
30
61
38
34
31
65
35
37
31
65
37
64
39
31
39
63
38
31
63
37
33
30
37
35
65
66
35
64
31
35
32
66
66
64
62
35
32
31
30
36
36
22
0a
64
65
70
65
6e
64
65
6e
63
69
65
73
20
3d
20
5b
0a
20
22
61
72
72
61
79
76
65
63
22
2c
0a
20
22
62
69
74
76
65
63
22
2c
0a
20
22
62
79
74
65
2d
73
6c
69
63
65
2d
63
61
73
74
22
2c
0a
20
22
69
6d
70
6c
2d
74
72
61
69
74
2d
66
6f
72
2d
74
75
70
6c
65
73
22
2c
0a
20
22
70
61
72
69
74
79
2d
73
63
61
6c
65
2d
63
6f
64
65
63
2d
64
65
72
69
76
65
22
2c
0a
20
22
73
65
72
64
65
22
2c
0a
5d
0a
0a
5b
5b
70
61
63
6b
61
67
65
5d
5d
0a
6e
61
6d
65
20
3d
20
22
70
61
72
69
74
79
2d
73
63
61
6c
65
2d
63
6f
64
65
63
2d
64
65
72
69
76
65
22
0a
76
65
72
73
69
6f
6e
20
3d
20
22
33
2e
31
2e
32
22
0a
73
6f
75
72
63
65
20
3d
20
22
72
65
67
69
73
74
72
79
2b
68
74
74
70
73
3a
2f
2f
67
69
74
68
75
62
2e
63
6f
6d
2f
72
75
73
74
2d
6c
61
6e
67
2f
63
72
61
74
65
73
2e
69
6f
2d
69
6e
64
65
78
22
0a
63
68
65
63
6b
73
75
6d
20
3d
20
22
63
34
35
65
64
31
66
33
39
37
30
39
66
35
61
38
39
33
33
38
66
61
62
35
30
65
35
39
38
31
36
62
32
65
38
38
31
35
66
35
62
62
35
38
32
37
36
65
37
64
64
66
39
61
66
64
34
39
35
66
37
33
66
38
22
0a
64
65
70
65
6e
64
65
6e
63
69
65
73
20
3d
20
5b
0a
20
22
70
72
6f
63
2d
6d
61
63
72
6f
2d
63
72
61
74
65
22
2c
0a
20
22
70
72
6f
63
2d
6d
61
63
72
6f
32
22
2c
0a
20
22
71
75
6f
74
65
22
2c
0a
20
22
73
79
6e
22
2c
0a
5d
0a
0a
5b
5b
70
61
63
6b
61
67
65
5d
5d
0a
6e
61
6d
65
20
3d
20
22
70
72
6f
63
2d
6d
61
63
72
6f
2d
63
72
61
74
65
22
0a
76
65
72
73
69
6f
6e
20
3d
20
22
31
2e
31
2e
33
22
0a
73
6f
75
72
63
65
20
3d
20
22
72
65
67
69
73
74
72
79
2b
68
74
74
70
73
3a
2f
2f
67
69
74
68
75
62
2e
63
6f
6d
2f
72
75
73
74
2d
6c
61
6e
67
2f
63
72
61
74
65
73
2e
69
6f
2d
69
6e
64
65
78
22
0a
63
68
65
63
6b
73
75
6d
20
3d
20
22
65
31
37
64
34
37
63
65
39
31
34
62
66
34
64
65
34
34
30
33
33
32
32
35
30
62
30
65
64
64
32
33
63
65
34
38
63
30
30
35
66
35
39
66
61
62
33
39
64
33
33
33
35
38
36
36
62
31
31
34
66
31
31
61
22
0a
64
65
70
65
6e
64
65
6e
63
69
65
73
20
3d
20
5b
0a
20
22
74
68
69
73
65
72
72
6f
72
22
2c
0a
20
22
74
6f
6d
6c
22
2c
0a
5d
0a
0a
5b
5b
70
61
63
6b
61
67
65
5d
5d
0a
6e
61
6d
65
20
3d
20
22
70
72
6f
63
2d
6d
61
63
72
6f
32
22
0a
76
65
72
73
69
6f
6e
20
3d
20
22
31
2e
30
2e
33
36
22
0a
73
6f
75
72
63
65
20
3d
20
22
72
65
67
69
73
74
72
79
2b
68
74
74
70
73
3a
2f
2f
67
69
74
68
75
62
2e
63
6f
6d
2f
72
75
73
74
2d
6c
61
6e
67
2f
63
72
61
74
65
73
2e
69
6f
2d
69
6e
64
65
78
22
0a
63
68
65
63
6b
73
75
6d
20
3d
20
22
63
37
33
34
32
64
35
38
38
33
66
62
63
63
61
65
31
63
63
33
37
61
32
33
35
33
62
30
39
63
38
37
63
39
62
30
66
33
61
66
64
37
33
66
35
66
62
39
62
62
61
36
38
37
61
31
66
37
33
33
62
30
32
39
22
0a
64
65
70
65
6e
64
65
6e
63
69
65
73
20
3d
20
5b
0a
20
22
75
6e
69
63
6f
64
65
2d
78
69
64
22
2c
0a
5d
0a
0a
5b
5b
70
61
63
6b
61
67
65
5d
5d
0a
6e
61
6d
65
20
3d
20
22
71
75
6f
74
65
22
0a
76
65
72
73
69
6f
6e
20
3d
20
22
31
2e
30
2e
31
37
22
0a
73
6f
75
72
63
65
20
3d
20
22
72
65
67
69
73
74
72
79
2b
68
74
74
70
73
3a
2f
2f
67
69
74
68
75
62
2e
63
6f
6d
2f
72
75
73
74
2d
6c
61
6e
67
2f
63
72
61
74
65
73
2e
69
6f
2d
69
6e
64
65
78
22
0a
63
68
65
63
6b
73
75
6d
20
3d
20
22
36
33
32
64
30
32
62
66
66
37
66
38
37
34
61
33
36
66
33
33
65
61
38
62
62
34
31
36
63
64
34
38
34
62
39
30
63
63
36
36
63
31
31
39
34
62
31
61
31
31
31
30
64
30
36
37
61
37
30
31
33
66
35
38
22
0a
64
65
70
65
6e
64
65
6e
63
69
65
73
20
3d
20
5b
0a
20
22
70
72
6f
63
2d
6d
61
63
72
6f
32
22
2c
0a
5d
0a
0a
5b
5b
70
61
63
6b
61
67
65
5d
5d
0a
6e
61
6d
65
20
3d
20
22
72
61
64
69
75
6d
22
0a
76
65
72
73
69
6f
6e
20
3d
20
22
30
2e
37
2e
30
22
0a
73
6f
75
72
63
65
20
3d
20
22
72
65
67
69
73
74
72
79
2b
68
74
74
70
73
3a
2f
2f
67
69
74
68
75
62
2e
63
6f
6d
2f
72
75
73
74
2d
6c
61
6e
67
2f
63
72
61
74
65
73
2e
69
6f
2d
69
6e
64
65
78
22
0a
63
68
65
63
6b
73
75
6d
20
3d
20
22
64
63
33
33
66
66
32
64
34
39
37
33
64
35
31
38
64
38
32
33
64
36
31
61
61
32
33
39
30
31
34
38
33
31
65
35
32
31
63
37
35
64
61
35
38
65
33
64
66
34
38
34
30
64
33
66
34
37
37
34
39
64
30
39
22
0a
0a
5b
5b
70
61
63
6b
61
67
65
5d
5d
0a
6e
61
6d
65
20
3d
20
22
73
63
61
6c
65
5f
66
69
78
74
75
72
65
73
22
0a
76
65
72
73
69
6f
6e
20
3d
20
22
30
2e
31
2e
30
22
0a
64
65
70
65
6e
64
65
6e
63
69
65
73
20
3d
20
5b
0a
20
22
69
74
65
72
74
6f
6f
6c
73
22
2c
0a
20
22
70
61
72
69
74
79
2d
73
63
61
6c
65
2d
63
6f
64
65
63
22
2c
0a
5d
0a
0a
5b
5b
70
61
63
6b
61
67
65
5d
5d
0a
6e
61
6d
65
20
3d
20
22
73
65
72
64
65
22
0a
76
65
72
73
69
6f
6e
20
3d
20
22
31
2e
30
2e
31
33
36
22
0a
73
6f
75
72
63
65
20
3d
20
22
72
65
67
69
73
74
72
79
2b
68
74
74
70
73
3a
2f
2f
67
69
74
68
75
62
2e
63
6f
6d
2f
72
75
73
74
2d
6c
61
6e
67
2f
63
72
61
74
65
73
2e
69
6f
2d
69
6e
64
65
78
22
0a
63
68
65
63
6b
73
75
6d
20
3d
20
22
63
65
33
31
65
32
34
62
30
31
65
31
65
35
32
34
64
66
39
36
66
31
63
32
66
64
64
30
35
34
34
30
35
66
38
64
37
33
37
36
32
34
39
61
35
31
31
30
38
38
36
66
62
34
62
36
35
38
34
38
34
37
38
39
22
0a
0a
5b
5b
70
61
63
6b
61
67
65
5d
5d
0a
6e
61
6d
65
20
3d
20
22
73
79
6e
22
0a
76
65
72
73
69
6f
6e
20
3d
20
22
31
2e
30
2e
38
39
22
0a
73
6f
75
72
63
65
20
3d
20
22
72
65
67
69
73
74
72
79
2b
68
74
74
70
73
3a
2f
2f
67
69
74
68
75
62
2e
63
6f
6d
2f
72
75
73
74
2d
6c
61
6e
67
2f
63
72
61
74
65
73
2e
69
6f
2d
69
6e
64
65
78
22
0a
63
68
65
63
6b
73
75
6d
20
3d
20
22
65
61
32
39
37
62
65
32
32
30
64
35
32
33
39
38
64
63
63
30
37
63
65
31
35
61
32
30
39
66
63
65
34
33
36
64
33
36
31
37
33
35
61
63
31
64
62
37
30
30
63
61
62
33
62
36
63
64
66
62
39
66
35
34
22
0a
64
65
70
65
6e
64
65
6e
63
69
65
73
20
3d
20
5b
0a
20
22
70
72
6f
63
2d
6d
61
63
72
6f
32
22
2c
0a
20
22
71
75
6f
74
65
22
2c
0a
20
22
75
6e
69
63
6f
64
65
2d
78
69
64
22
2c
0a
5d
0a
0a
5b
5b
70
61
63
6b
61
67
65
5d
5d
0a
6e
61
6d
65
20
3d
20
22
74
61
70
22
0a
76
65
72
73
69
6f
6e
20
3d
20
22
31
2e
30
2e
31
22
0a
73
6f
75
72
63
65
20
3d
20
22
72
65
67
69
73
74
72
79
2b
68
74
74
70
73
3a
2f
2f
67
69
74
68
75
62
2e
63
6f
6d
2f
72
75
73
74
2d
6c
61
6e
67
2f
63
72
61
74
65
73
2e
69
6f
2d
69
6e
64
65
78
22
0a
63
68
65
63
6b
73
75
6d
20
3d
20
22
35
35
39
33
37
65
31
37
39
39
31
38
35
62
31
32
38
36
33
64
34
34
37
66
34
32
35
39
37
65
64
36
39
64
39
39
32
38
36
38
36
62
38
64
38
38
61
31
64
66
31
37
33
37
36
61
30
39
37
64
38
33
36
39
22
0a
0a
5b
5b
70
61
63
6b
61
67
65
5d
5d
0a
6e
61
6d
65
20
3d
20
22
74
68
69
73
65
72
72
6f
72
22
0a
76
65
72
73
69
6f
6e
20
3d
20
22
31
2e
30
2e
33
30
22
0a
73
6f
75
72
63
65
20
3d
20
22
72
65
67
69
73
74
72
79
2b
68
74
74
70
73
3a
2f
2f
67
69
74
68
75
62
2e
63
6f
6d
2f
72
75
73
74
2d
6c
61
6e
67
2f
63
72
61
74
65
73
2e
69
6f
2d
69
6e
64
65
78
22
0a
63
68
65
63
6b
73
75
6d
20
3d
20
22
38
35
34
62
61
62
65
35
32
65
34
64
66
31
36
35
33
37
30
36
62
39
38
66
63
66
63
30
35
38
34
33
30
31
30
30
33
39
62
34
30
36
38
37
35
39
33
30
61
37
30
65
34
64
39
36
34
34
65
35
63
34
31
37
22
0a
64
65
70
65
6e
64
65
6e
63
69
65
73
20
3d
20
5b
0a
20
22
74
68
69
73
65
72
72
6f
72
2d
69
6d
70
6c
22
2c
0a
5d
0a
0a
5b
5b
70
61
63
6b
61
67
65
5d
5d
0a
6e
61
6d
65
20
3d
20
22
74
68
69
73
65
72
72
6f
72
2d
69
6d
70
6c
22
0a
76
65
72
73
69
6f
6e
20
3d
20
22
31
2e
30
2e
33
30
22
0a
73
6f
75
72
63
65
20
3d
20
22
72
65
67
69
73
74
72
79
2b
68
74
74
70
73
3a
2f
2f
67
69
74
68
75
62
2e
63
6f
6d
2f
72
75
73
74
2d
6c
61
6e
67
2f
63
72
61
74
65
73
2e
69
6f
2d
69
6e
64
65
78
22
0a
63
68
65
63
6b
73
75
6d
20
3d
20
22
61
61
33
32
66
64
33
66
36
32
37
66
33
36
37
66
65
31
36
66
38
39
33
65
32
35
39
37
61
65
33
63
30
35
30
32
30
66
38
62
62
61
32
36
36
36
61
34
65
36
65
61
37
33
64
33
37
37
65
35
37
31
34
62
22
0a
64
65
70
65
6e
64
65
6e
63
69
65
73
20
3d
20
5b
0a
20
22
70
72
6f
63
2d
6d
61
63
72
6f
32
22
2c
0a
20
22
71
75
6f
74
65
22
2c
0a
20
22
73
79
6e
22
2c
0a
5d
0a
0a
5b
5b
70
61
63
6b
61
67
65
5d
5d
0a
6e
61
6d
65
20
3d
20
22
74
6f
6d
6c
22
0a
76
65
72
73
69
6f
6e
20
3d
20
22
30
2e
35
2e
38
22
0a
73
6f
75
72
63
65
20
3d
20
22
72
65
67
69
73
74
72
79
2b
68
74
74
70
73
3a
2f
2f
67
69
74
68
75
62
2e
63
6f
6d
2f
72
75
73
74
2d
6c
61
6e
67
2f
63
72
61
74
65
73
2e
69
6f
2d
69
6e
64
65
78
22
0a
63
68
65
63
6b
73
75
6d
20
3d
20
22
61
33
31
31
34
32
39
37
30
38
32
36
37
33
33
64
66
38
32
34
31
65
66
33
35
64
63
30
34
30
65
66
39
38
63
36
37
39
61
62
31
34
64
37
63
33
65
35
34
64
38
32
37
30
39
39
62
33
61
63
65
63
61
61
22
0a
64
65
70
65
6e
64
65
6e
63
69
65
73
20
3d
20
5b
0a
20
22
73
65
72
64
65
22
2c
0a
5d
0a
0a
5b
5b
70
61
63
6b
61
67
65
5d
5d
0a
6e
61
6d
65
20
3d
20
22
75
6e
69
63
6f
64
65
2d
78
69
64
22
0a
76
65
72
73
69
6f
6e
20
3d
20
22
30
2e
32
2e
32
22
0a
73
6f
75
72
63
65
20
3d
20
22
72
65
67
69
73
74
72
79
2b
68
74
74
70
73
3a
2f
2f
67
69
74
68
75
62
2e
63
6f
6d
2f
72
75
73
74
2d
6c
61
6e
67
2f
63
72
61
74
65
73
2e
69
6f
2d
69
6e
64
65
78
22
0a
63
68
65
63
6b
73
75
6d
20
3d
20
22
38
63
63
62
38
32
64
36
31
66
38
30
61
36
36
33
65
66
65
31
66
37
38
37
61
35
31
62
31
36
62
35
61
35
31
65
33
33
31
34
64
36
61
63
33
36
35
62
30
38
36
33
39
66
35
32
33
38
37
62
33
33
66
33
22
0a
0a
5b
5b
70
61
63
6b
61
67
65
5d
5d
0a
6e
61
6d
65
20
3d
20
22
77
79
7a
22
0a
76
65
72
73
69
6f
6e
20
3d
20
22
30
2e
35
2e
30
22
0a
73
6f
75
72
63
65
20
3d
20
22
72
65
67
69
73
74
72
79
2b
68
74
74
70
73
3a
2f
2f
67
69
74
68
75
62
2e
63
6f
6d
2f
72
75
73
74
2d
6c
61
6e
67
2f
63
72
61
74
65
73
2e
69
6f
2d
69
6e
64
65
78
22
0a
63
68
65
63
6b
73
75
6d
20
3d
20
22
33
30
62
33
31
35
39
34
66
32
39
64
32
37
30
33
36
63
33
38
33
62
35
33
62
35
39
65
64
33
34
37
36
38
37
34
64
35
31
38
66
30
65
66
62
31
35
31
62
32
37
61
34
63
32
37
35
31
34
31
33
39
30
65
22
0a
64
65
70
65
6e
64
65
6e
63
69
65
73
20
3d
20
5b
0a
20
22
74
61
70
22
2c
0a
5d
0a
`;
